src/marker.rs

```diff
@@ -528,21 +286,133 @@ impl<'py> Python<'py> {
+        use std::time::Duration;
+
+        use parking_lot::{const_mutex, Mutex};
+
+        use crate::impl_::panic::PanicTrap;
+
         // Use a guard pattern to handle reacquiring the GIL,
         // so that the GIL will be reacquired even if `f` panics.
         // The `Send` bound on the closure prevents the user from
         // transferring the `Python` token into the closure.
         let _guard = unsafe { SuspendGIL::new() };
-        f()
+
+        // To close soundness loopholes w.r.t. `send_wrapper` or `scoped-tls`,
+        // we run the closure on a separate thread so that it cannot
+        // access thread-local storage from the current thread.
+
+        // 1. Construct a task
+        struct Task(*mut dyn FnMut());
+        unsafe impl Send for Task {}
+
+        let (result_sender, result_receiver) = sync_channel::<Result<T>>(0);
+
+        let mut f = Some(f);
+
+        let mut task = || {
+            let f = f.take().unwrap();
+
+            let result = catch_unwind(AssertUnwindSafe(f));
+
+            result_sender.send(result).unwrap();
+        };
+
+        // SAFETY: the current thread will block until the closure has returned
+        let task = Task(unsafe { transmute(&mut task as &mut dyn FnMut()) });
+
+        // 2. Dispatch task to waiting thread, spawn new thread if necessary
+        let trap = PanicTrap::new(
+            "allow_threads panicked while stack data was accessed by another thread which is a bug",
+        );
+
+        static THREADS: Mutex<Vec<SyncSender<Task>>> = const_mutex(Vec::new());
+
+        enum State {
+            Pending(Task),
+            Dispatched(SyncSender<Task>),
+        }
+
+        let mut state = State::Pending(task);
+
+        while let Some(task_sender) = THREADS.lock().pop() {
+            match state {
+                State::Pending(task) => match task_sender.send(task) {
+                    Ok(()) => {
+                        state = State::Dispatched(task_sender);
+                        break;
+                    }
+                    Err(SendError(task)) => {
+                        state = State::Pending(task);
+                        continue;
+                    }
+                },
+                State::Dispatched(_task_sender) => unreachable!(),
+            }
+        }
+
+        let task_sender = match state {
+            State::Pending(task) => {
+                let (task_sender, task_receiver) = sync_channel::<Task>(0);
+
+                spawn(move || {
+                    while let Ok(task) = task_receiver.recv_timeout(Duration::from_secs(60)) {
+                        // SAFETY: all data accessed by `task` will stay alive until it completes
+                        unsafe { (*task.0)() };
+                    }
+                });
+
+                task_sender.send(task).unwrap();
+
+                task_sender
+            }
+            State::Dispatched(task_sender) => task_sender,
+        };
```

Comment:
Will drop `enum State` even though I have to say that I did the like explicitness of turning that state into types instead of handling it via control flow, but the named block is indeed more succinct.

I would however prefer to keep moving `task` instead of deriving `Copy` or doing the basically equivalent `Task(task.0)` dance because having only one handle by which the task is called is an important safety invariant here.

Reply:
Yes I didn't love the `Task(task.0)` dance but missed earlier a way to avoid that, looks like this can work:

```rust
        let mut task = task;
        // Try to get an idle worker thread
        while let Some(task_sender) = THREADS.pop() {
            task = match task_sender.send(task) {
                Ok(()) => break 'sender task_sender,
                // the worker thread timed out and shutdown
                Err(SendError(task)) => task,
            }
        };
```

Reply:
Hhhmm, I just bit the bullet and used a new name for `SendError(same_task) => task = same_task,` since that seem more readable to me even if I hate having to name things just for the syntactical sake of it.